jbpm-runtime-manager/src/main/java/org/jbpm/runtime/manager/impl/lock/AbstractRuntimeManagerLockStrategy.java

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright 2020 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.jbpm.runtime.manager.impl.lock;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+import org.jbpm.runtime.manager.spi.RuntimeManagerLock;
+import org.jbpm.runtime.manager.spi.RuntimeManagerLockFactory;
+import org.jbpm.runtime.manager.spi.RuntimeManagerLockStrategy;
+import org.kie.api.runtime.manager.RuntimeEngine;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public abstract class AbstractRuntimeManagerLockStrategy implements RuntimeManagerLockStrategy {
+
+    protected static final Logger logger = LoggerFactory.getLogger(AbstractRuntimeManagerLockStrategy.class);
+
+    protected Map<Long, RuntimeManagerLockThreadsInfo> engineLocks; 
+
+    private RuntimeManagerLockFactory runtimeManagerLockFactory;
+
+    
+    public AbstractRuntimeManagerLockStrategy() {
+        this.runtimeManagerLockFactory = new DefaultRuntimeManagerLockFactory();
+        this.engineLocks = new HashMap<>();
+    }
+
+    @Override
+    public void init(RuntimeManagerLockFactory factory) {
+        this.runtimeManagerLockFactory = factory;
+    }
+
+    @Override
+    public void lock(Long id, RuntimeEngine runtime) throws InterruptedException {
+        RuntimeManagerLockThreadsInfo lockThreadsInfo = null;
+        synchronized (engineLocks) {
+            if(!engineLocks.containsKey(id)) {
+                engineLocks.put(id, new RuntimeManagerLockThreadsInfo(runtimeManagerLockFactory.newRuntimeManagerLock()));
+            }
+            lockThreadsInfo = engineLocks.get(id);
+            lockThreadsInfo.set();
+        }
+        logger.debug("Trying to get a lock {} for {} by {}", lockThreadsInfo, id, runtime);
+        try {
+            lock(lockThreadsInfo.getRuntimeManagerLock());
+        } catch(InterruptedException e) {
+            logger.warn("Interrupted lock {}", lockThreadsInfo.getRuntimeManagerLock());
+            throw e;
+        }
+        logger.debug("Lock {} taken for {} by {} for waiting threads by {}", lockThreadsInfo, id, runtime, lockThreadsInfo.hasQueuedThreads());
+    }
+
+    protected abstract void lock(RuntimeManagerLock lock) throws InterruptedException;
+    
+    @Override
+    public void unlock(Long id, RuntimeEngine runtime) {
```

Comment:
Check first if id is null to return

Reply:
id won't be null ... it is guaranteed at runtime manager level